tensorflow_recommenders_addons/dynamic_embedding/python/kernel_tests/horovod_sync_train_test.py

```diff
@@ -0,0 +1,105 @@
+# limitations under the License.
+# ==============================================================================
+"""unit tests of dynamic embedding optimizer ops
+"""
+from __future__ import absolute_import
+from __future__ import division
+from __future__ import print_function
+
+import itertools
+import pytest
+import tensorflow as tf
+import horovod.tensorflow as hvd
+
+from tensorflow_recommenders_addons import dynamic_embedding as de
+
+from tensorflow.core.protobuf import config_pb2
+from tensorflow.python.framework import dtypes
+from tensorflow.python.framework import test_util
+from tensorflow.python.ops import math_ops
+from tensorflow.python.ops import variables
+from tensorflow.python.platform import test
+from tensorflow.python.training import adam
+from tensorflow.python.training import monitored_session
+from tensorflow.python.training import training_util
+
+default_config = config_pb2.ConfigProto(
+    allow_soft_placement=True,
+    gpu_options=config_pb2.GPUOptions(allow_growth=True))
+
+
+class HorovodTest(test.TestCase):
+
+  @test_util.deprecated_graph_mode_only
+  def test_adam_minimize_trainable(self):
+    base_opt = adam.AdamOptimizer(1.0)
+    test_opt = adam.AdamOptimizer(1.0)
+    self.common_minimize_trainable(base_opt, test_opt, name="adam")
+
+  def common_minimize_trainable(self, base_opt, test_opt, name):
+    tf.config.set_soft_device_placement(True)
+    hvd.init()
+    base_opt = de.DynamicEmbeddingOptimizer(base_opt, horovod_synchronous=True)
+    for dtype, run_step, dim in itertools.product([dtypes.float32], [1], [10]):
+      x = tf.random.uniform(shape=[32, dim])
+      y = tf.zeros([32, 1])
+
+      global_step = training_util.create_global_step()
+
+      base_weight = tf.compat.v1.get_variable(name="base_weights",
+                                              initializer=tf.ones([10, 1]))
+
+      base_logits = tf.nn.relu(math_ops.matmul(x, base_weight))
+      base_loss = tf.nn.sigmoid_cross_entropy_with_logits(labels=y,
+                                                          logits=base_logits)
+
+      base_opt_op = base_opt.minimize(base_loss,
+                                      global_step,
+                                      var_list=[base_weight])
+
+      test_weight = tf.compat.v1.get_variable(name="test_weights",
+                                              initializer=tf.ones([10, 1]))
+
+      test_logits = tf.nn.relu(math_ops.matmul(x, test_weight))
+      test_loss = tf.nn.sigmoid_cross_entropy_with_logits(labels=y,
+                                                          logits=test_logits)
+
+      grads_and_vars = test_opt.compute_gradients(test_loss,
+                                                  var_list=[test_weight])
+      var_list = []
+      aggregated_grad = []
+      for grad, var in grads_and_vars:
+        var_list.append(var)
+        aggregated_grad.append(hvd.allreduce(grad, op=hvd.Sum))
```

Comment:
And just for discussion, I noticed that Horovod implements lots of features and optimizations on its DistributedOptimizer like tensor fusion, grouped allreduce, adasum, gradient compression, etc.., in the latest-nth versions. Use the hvd.allreduce API may not reuse these features.